torch/fx/experimental/proxy_tensor.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import functools
+from typing import Any, Dict, Optional, Tuple, Callable, Union
+import torch
+from torch._C import _disabled_torch_function_impl
+import torch.utils._pytree as pytree
+from torch.fx import Tracer, GraphModule
+import torch.fx as fx
+from torch.fx.passes.shape_prop import _extract_tensor_metadata
+from contextlib import contextmanager
+
+aten = torch.ops.aten
+
+CURRENT_DECOMPOSITION_TABLE = {}
+
+
+@contextmanager
+def no_dispatch():
+    guard = torch._C._DisableTorchDispatch()
+    try:
+        yield
+    finally:
+        del guard
+
+
+@contextmanager
+def decompose(decomposition_table):
+    global CURRENT_DECOMPOSITION_TABLE
+    old_decomposition_table = CURRENT_DECOMPOSITION_TABLE
+    CURRENT_DECOMPOSITION_TABLE = decomposition_table
+    try:
+        yield CURRENT_DECOMPOSITION_TABLE
+    finally:
+        CURRENT_DECOMPOSITION_TABLE = old_decomposition_table
```

Comment:
@Chillee do you think this should be the done and dusted API for decomps in AOTAutograd? I now agree that it should be possible to directly program extra decomps as part of the tracing process, but I still want control of the default set of decompositions (which we're gonna put in PyTorch core) to be based on what the backend declares it supports.

I suppose one could do this compositionally in the current API with something like this: `decompose({**get_decomps(dont_decompose), **custom_decomps})`. (this inverts the sense in which `get_decomps` is currently programmed; instead of asking for decomps, you say everything you DON'T want decomposed.) But this seems like a weird way to do the API if you're talking in terms of building an array of decompositions; it seems like maybe this should be abstracted away into a higher level API

Reply:
I think in practice... backends like NVFuser don't actually necessarily want to "try and decompose everything they don't support". For example, something like `slice_backward` decomposes into a `new_zeros` and a `slice_scatter` call. Now, NVFuser supports neither of these new ops, and since we've decomposed them they're now slower.

Essentially, one nice thing about this API is that the "default option" (i.e. decompose nothing) is very easy to reason about. Similarly, if you're only decomposing a few ops, it's very easy to reason about it.

I could be convinced that there could be a better API though.

Reply:
OK, agreed this is not good enough.

It seems like something better might be if we had metadata saying what a decomposition would decompose to. Then instead NVFuser can define "this is the stuff that I actually understand", and then we only do decompositions that (eventually) produce things we understand (but this is not 100% well specified, because what if a decomp produces some understandable stuff and some not understandable stuff, is it still profitable to decompose).